File: LibroClubSoft/LibroClubSoftExternoController/src/pe/edu/pucp/LibroClubSoft/externo/mysql/UsuarioMySQL.java
package pe.edu.pucp.LibroClubSoft.externo.mysql;

import pe.edu.pucp.LibroClubSoft.externo.dao.UsuarioDAO;

import java.sql.CallableStatement;
import java.sql.Connection;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import pe.edu.pucp.LibroClubSoft.config.DBManager;
import pe.edu.pucp.LibroClubSoft.externo.model.TipoMembresia;
import pe.edu.pucp.LibroClubSoft.externo.model.Usuario;

public class UsuarioMySQL implements UsuarioDAO {

    private Connection con;
    private ResultSet rs;
    private CallableStatement cs;

    @Override
    public int insertar(Usuario obj) {
        int resultado = 0;
        System.out.println(obj.getDni());
        System.out.println(obj.getNombre());
        System.out.println(obj.getFecha_nacimiento());
        System.out.println(obj.getSexo());
        System.out.println(obj.getTelefono());
        System.out.println(obj.getMembresia());
        try {
            con = DBManager.getInstance().getConnection();
            cs = con.prepareCall("{call USUARIO_INSERTAR(?,?,?,?,?,?,?)}");

            cs.registerOutParameter("_id_persona", java.sql.Types.INTEGER);
            cs.setString("_dni", obj.getDni());
            cs.setString("_nombre", obj.getNombre());
            cs.setDate("_fecha_nacimiento", new java.sql.Date(obj.getFecha_nacimiento().getTime()));
            cs.setString("_genero", String.valueOf(obj.getSexo()));
            cs.setInt("_telefono", obj.getTelefono());

            String m = obj.getMembresia().toString();
            int id_membresia;

            switch (m) {
                case "Ninguna":
                    id_membresia = 0;
                    System.out.println("NINGUNA");
                    break;
                case "Plata":
                    id_membresia = 1;
                    System.out.println("PLATA");
                    break;
                case "Oro":
                    id_membresia = 2;
                    System.out.println("ORO");
                    break;
                default:
                    // En caso de que el estado no coincida con ninguno de los casos anteriores
                    // Se podría lanzar una excepción, asignar un valor por defecto, o manejarlo de acuerdo a tu lógica de negocio
                    // Aquí simplemente se asigna un valor negativo para indicar un estado no reconocido
                    id_membresia = -1;
                    break;
            }
            cs.setInt("_fid_tipo_membresia", id_membresia);

            cs.executeUpdate();
            obj.setId_persona(cs.getInt("_id_persona"));
            resultado = obj.getId_persona();
            con.close();
        } catch (SQLException ex) {
            System.out.println(ex.getMessage());
        }
        return resultado;
    }

    @Override
    public ArrayList<Usuario> listar() {
        ArrayList<Usuario> usuarios = new ArrayList<>();

        try {
            con = DBManager.getInstance().getConnection();
            cs = con.prepareCall("{call USUARIO_LISTAR()}");
            rs = cs.executeQuery();
            while (rs.next()) {
                Usuario usuario = new Usuario();
                usuario.setId_persona(rs.getInt("id_usuario"));
                usuario.setDni(rs.getString("dni"));
                usuario.setNombre(rs.getString("nombre"));
                usuario.setFecha_nacimiento(rs.getDate("fecha_nacimiento"));
                usuario.setSexo(rs.getString("genero").charAt(0));
                usuario.setTelefono(rs.getInt("telefono"));
                System.out.println(rs.getString("tipo_membresia"));
                usuario.setMembresia(TipoMembresia.valueOf(rs.getString("tipo_membresia")));

                usuarios.add(usuario);
            }
        } catch (SQLException ex) {
            System.out.println(ex.getMessage());
        } finally {
            try {
                con.close();
            } catch (SQLException ex) {
                System.out.println(ex.getMessage());
            }
        }
        return usuarios;
    }
    
    @Override
    public ArrayList<Usuario> listarMembresia(int id) {
        ArrayList<Usuario> usuarios = new ArrayList<>();

        try {
            con = DBManager.getInstance().getConnection();
            cs = con.prepareCall("{call USUARIO_LISTAR_X_MEMBRESIA(?)}");
            cs.setInt("_id_tipo_membresia", id);
            rs = cs.executeQuery();
            while (rs.next()) {
                Usuario usuario = new Usuario();
                usuario.setId_persona(rs.getInt("id_usuario"));
                usuario.setDni(rs.getString("dni"));
                usuario.setNombre(rs.getString("nombre"));
                usuario.setFecha_nacimiento(rs.getDate("fecha_nacimiento"));
                usuario.setSexo(rs.getString("genero").charAt(0));
                usuario.setTelefono(rs.getInt("telefono"));
                usuario.setMembresia(TipoMembresia.valueOf(rs.getString("tipo_membresia")));

                usuarios.add(usuario);
            }
        } catch (SQLException ex) {
            System.out.println(ex.getMessage());
        } finally {
            try {
                con.close();
            } catch (SQLException ex) {
                System.out.println(ex.getMessage());
            }
        }
        return usuarios;
    }
    
    @Override
    public Usuario ObtenerDni(String dni) {
        Usuario usuario = new Usuario();

        try {
            con = DBManager.getInstance().getConnection();
            cs = con.prepareCall("{call USUARIO_OBTENER_X_DNI(?)}");
            cs.setString("_dni", dni);
            rs = cs.executeQuery();
            if (rs.next()) {
                usuario.setId_persona(rs.getInt("id_usuario"));
                usuario.setDni(rs.getString("dni"));
                usuario.setNombre(rs.getString("nombre"));
                usuario.setFecha_nacimiento(rs.getDate("fecha_nacimiento"));
                usuario.setSexo(rs.getString("genero").charAt(0));
                usuario.setTelefono(rs.getInt("telefono"));
                usuario.setMembresia(TipoMembresia.valueOf(rs.getString("tipo_membresia")));
            }
            cs.close();
        } catch (SQLException ex) {
            System.out.println(ex.getMessage());
        } finally {
            try {
                con.close();
            } catch (SQLException ex) {
                System.out.println(ex.getMessage());
            }
        }
        return usuario;
    }

    @Override
    public int eliminar(int id) {
        int resultado = 0;
        try {
            con = DBManager.getInstance().getConnection();
            cs = con.prepareCall("{call USUARIO_ELIMINAR(?)}");
            cs.setInt("_id_persona", id);

            resultado = cs.executeUpdate();
        } catch (SQLException ex) {
            System.out.println(ex.getMessage());
        } finally {
            try {
                con.close();
            } catch (SQLException ex) {
                System.out.println(ex.getMessage());
            }
        }
        return resultado;
    }

    @Override
    public int modificar(Usuario obj) {
        int resultado = 0;
        try {
            con = DBManager.getInstance().getConnection();

            cs = con.prepareCall("{call USUARIO_MODIFICAR(?,?,?,?,?,?,?)}");
            cs.setInt("_id_persona", obj.getId_persona());
            cs.setDate("_fecha_nacimiento", new java.sql.Date(obj.getFecha_nacimiento().getTime()));
            cs.setString("_dni", obj.getDni());
            cs.setString("_nombre", obj.getNombre());
            cs.setString("_genero", String.valueOf(obj.getSexo()));
            cs.setInt("_telefono", obj.getTelefono());
//            cs.setDate("_fecha_afiliacion", new java.sql.Date(obj.getFecha_afiliacion().getTime()));

            String m = obj.getMembresia().toString();
            int id_membresia;

            switch (m) {
                case "Ninguna":
                    id_membresia = 0;
                    System.out.println("NINGUNA ");
                    break;
                case "Plata":
                    id_membresia = 1;
                    System.out.println("PLARTA ");
                    break;
                case "Oro":
                    id_membresia = 2;
                    System.out.println("ORO ");
                    break;
                default:
                    // En caso de que el estado no coincida con ninguno de los casos anteriores
                    // Se podría lanzar una excepción, asignar un valor por defecto, o manejarlo de acuerdo a tu lógica de negocio
                    // Aquí simplemente se asigna un valor negativo para indicar un estado no reconocido
                    id_membresia = -1;
                    break;
            }

            cs.setInt("_fid_tipo_membresia", id_membresia);

            resultado = cs.executeUpdate();
            cs.close();
        } catch (SQLException ex) {
            System.out.println(ex.getMessage());
        }
        return resultado;
    }

    @Override
    public ArrayList<Usuario> listarPorNombre(String cadena){
        ArrayList<Usuario> usuarios =  new ArrayList<>();
        try{
            
            con = DBManager.getInstance().getConnection();
            cs = con.prepareCall("{call USUARIO_LISTAR_NOMBRE(?)}");
            cs.setString("_nombre",cadena);
            rs = cs.executeQuery();
            while(rs.next()){
                Usuario usuario = new Usuario();
                usuario.setId_persona(rs.getInt("fid_usuario"));
                usuario.setNombre(rs.getString("nombre"));
                usuario.setDni(rs.getString("dni"));
                usuario.setMembresia(TipoMembresia.valueOf(rs.getString("tipo_membresia")));
                
                
                usuarios.add(usuario);
            }
        }catch(Exception ex){
            System.out.println(ex.getMessage());
        }finally{
            try{con.close();}catch(SQLException ex){System.out.println(ex.getMessage());}
        }
        return usuarios;
    }
}

File: LibroClubViews/LibroClub/LibroClubController/MySQL/UsuarioMySQL.cs
using MySql.Data.MySqlClient;
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using LibroClubController.DAO;
using LibroClubModels;
using static Mysqlx.Crud.Order.Types;

namespace LibroClubController.MySQL
{
	public class UsuarioMySQL: UsuarioDAO
	{
		private MySqlConnection con;
		private MySqlCommand comando;
		private MySqlDataReader lector;
		public int insertar(Usuario usuario)
		{
			int resultado = 0;

			try
			{
				con = DBManager.Instance.Connection;
				//con.Open();
				comando = new MySqlCommand();
				comando.Connection = con;
				comando.CommandType = CommandType.StoredProcedure;
				comando.CommandText = "USUARIO_INSERTAR";

				comando.Parameters.Add("_id_persona", MySqlDbType.Int32).Direction = ParameterDirection.Output;
				//aqui se especifica que "_id_pelicula" va ser un Int32, para luego asignarlo como un output en el SQL

				comando.Parameters.AddWithValue("_dni", usuario.DNI);
				comando.Parameters.AddWithValue("_nombre", usuario.Nombre);
				comando.Parameters.AddWithValue("_fecha_nacimiento", usuario.Fecha_Nacimiento);
				comando.Parameters.AddWithValue("_genero", usuario.Genero);
				comando.Parameters.AddWithValue("_telefono", usuario.Telefono);
				comando.Parameters.AddWithValue("_fid_tipo_membresia", usuario.Id_Membresia);

				resultado = comando.ExecuteNonQuery();

				usuario.ID_Usuario = Int32.Parse(comando.Parameters["_id_persona"].Value.ToString());

			}
			catch (Exception ex)
			{
				throw new Exception(ex.Message);
			}
			finally
			{
				try { con.Close(); } catch (Exception ex) { throw new Exception(ex.Message); }
			}
			return resultado;
		}

		public List<Usuario> listar()
		{
			List<Usuario> usuarios = new List<Usuario>();

			try
			{
				con = DBManager.Instance.Connection;
				//con.Open();
				comando = new MySqlCommand();
				comando.Connection = con;
				comando.CommandType = System.Data.CommandType.StoredProcedure;
				comando.CommandText = "USUARIO_LISTAR";//name in Script SQL
				lector = comando.ExecuteReader();
				while (lector.Read())
				{
					Usuario usuario = new Usuario();
					usuario.ID_Usuario = lector.GetInt32("id_usuario");

					usuario.Persona = new Persona();
					usuario.Persona.DNI = lector.GetString("dni");
					usuario.Persona.Nombre = lector.GetString("nombre");
					usuario.Persona.Fecha_Nacimiento = lector.GetDateTime("fecha_nacimiento");
					usuario.Persona.Genero = lector.GetChar("genero");
					usuario.Persona.Telefono = lector.GetInt32("telefono");
					
					usuario.TipoMembresia = new TipoMembresia();
					usuario.TipoMembresia.Nombre = lector.GetString("tipo_membresia");

					usuarios.Add(usuario);
				}

			}
			catch (Exception ex)
			{
				throw new Exception(ex.Message);
			}
			finally
			{
				try { con.Close(); } catch (Exception ex) { throw new Exception(ex.Message); }
			}
			return usuarios;
		}

		public int eliminar(int id)
		{
			int resultado = 0;
			try
			{
				con = DBManager.Instance.Connection;
				//con.Open();
				comando = new MySqlCommand();
				comando.Connection = con;
				comando.CommandType = CommandType.StoredProcedure;
				comando.CommandText = "USUARIO_ELIMINAR";

				comando.Parameters.AddWithValue("_id_persona", id);

				resultado = comando.ExecuteNonQuery();

			}
			catch (Exception ex)
			{
				throw new Exception(ex.Message);
			}
			finally
			{
				try { con.Close(); } catch (Exception ex) { throw new Exception(ex.Message); }
			}
			return resultado;
		}

		public int modificar(int id, String dni, String nombre, DateTime fecha_nacimiento,
			char genero, int telefono, int tipo_membresia)
		{
			int resultado = 0;
			try
			{
				con = DBManager.Instance.Connection;
				//con.Open();
				comando = new MySqlCommand();
				comando.Connection = con;
				comando.CommandType = CommandType.StoredProcedure;
				comando.CommandText = "USUARIO_MODIFICAR";

				comando.Parameters.AddWithValue("_id_persona", id);
				comando.Parameters.AddWithValue("_dni", dni);
				comando.Parameters.AddWithValue("_nombre", nombre);
				comando.Parameters.AddWithValue("_fecha_nacimiento", fecha_nacimiento);
				comando.Parameters.AddWithValue("_genero", genero);
				comando.Parameters.AddWithValue("_telefono", telefono);
				comando.Parameters.AddWithValue("_fid_tipo_membresia", tipo_membresia);

				resultado = comando.ExecuteNonQuery();

			}
			catch (Exception ex)
			{
				throw new Exception(ex.Message);
			}
			finally
			{
				try { con.Close(); } catch (Exception ex) { throw new Exception(ex.Message); }
			}
			return resultado;
		}
	}
}
